nemo/collections/llm/quantization/quantizer.py

```diff
@@ -0,0 +1,332 @@
+
+from nemo import lightning as nl
+from nemo.collections import llm
+from nemo.collections.nlp.models.language_modeling.megatron.gpt_layer_modelopt_spec import get_gpt_layer_modelopt_spec
+from nemo.collections.nlp.parts.utils_funcs import torch_dtype_from_precision
+from nemo.utils import logging
+
+try:
+    import modelopt.torch.quantization as mtq
+    from modelopt.torch.export import export_tensorrt_llm_checkpoint
+
+    QUANT_CFG_CHOICES = {
+        "int8": mtq.INT8_DEFAULT_CFG,
+        "int8_sq": mtq.INT8_SMOOTHQUANT_CFG,
+        "fp8": mtq.FP8_DEFAULT_CFG,
+        "int4_awq": mtq.INT4_AWQ_CFG,
+        "w4a8_awq": mtq.W4A8_AWQ_BETA_CFG,
+        "int4": mtq.INT4_BLOCKWISE_WEIGHT_ONLY_CFG,
+    }
+
+    HAVE_MODELOPT = True
+
+except (ImportError, ModuleNotFoundError) as e:
+    HAVE_MODELOPT = False
+    HAVE_MODELOPT_ERROR = e
+
+
+SUPPORTED_DTYPE = [16, "16", "bf16"]  # Default precision for non-quantized layers
+
+
+def get_modelopt_decoder_type(config: llm.GPTConfig) -> str:
+    """Infers the modelopt decoder type from GPTConfig class"""
+    mapping = [
+        (llm.Baichuan2Config, "baichuan"),
+        (llm.ChatGLMConfig, "chatglm"),
+        (llm.GemmaConfig, "gemma"),
+        (llm.LlamaConfig, "llama"),
+        (llm.MistralConfig7B, "llama"),
+        (llm.MixtralConfig, "llama"),
+        (llm.NemotronConfig, "gptnext"),
+        (llm.Qwen2Config, "qwen"),
+        # TODO: (llm.StarcoderConfig,   ""),
+        (llm.Starcoder2Config, "gptnext"),
+    ]
+
+    for config_class, decoder_type in mapping:
+        if isinstance(config, config_class):
+            return decoder_type
+
+    logging.warning("Could not directly infer the decoder type")
+    # TODO: Add a reasonable behavior for GPTConfig (for instance based on position_embedding_type)
+    return "llama"
+
+
+# TODO: Support PP
+class Quantizer:
+    """Post-training quantization (PTQ) and TRT-LLM export of Nemo checkpoints.
+
+    PTQ converts selected model layers to low-precision format (e.g., INT4, FP8) for efficient serving.
+    The process consist of several steps:
+
+        1. Loading a Nemo model from disk using appropriate parallelism strategy
+        2. Calibrating the model to obtain appropriate algorithm-specific scaling factors
+        3. Producing output directory
+
+    The output directory produced is intended to be consumed by TensorRT-LLM toolbox
+    for efficient inference. This can be achieved using NeMo inference containers.
+
+    Available quantization methods are listed in `QUANT_CFG_CHOICES` dictionary above.
+    Please consult Model Optimizer documentation https://nvidia.github.io/TensorRT-Model-Optimizer/ for details.
+
+    Quantization algorithm can also be conveniently set to None to perform only weights export step
+    for TensorRT-LLM deployment. This is useful to getting baseline results for a full-precision model.
+    """
+
+    def __init__(self, quantization_config: dict, export_config: dict):
```

Comment:
We would like to move away from dictionaries everywhere, please turn this into dataclasses.

Reply:
Done, I've added `QuantizationConfig` and `ExportConfig` dataclasses